IndyAV/simulation/indyav_gazebo/src/backwheel_plugin.cpp

```diff
@@ -10,29 +10,37 @@ void BackWheelPlugin::Load(gazebo::physics::ModelPtr _model, sdf::ElementPtr _sd
 {
   WheelPlugin::Load(_model, _sdf);
 
-  GZ_ASSERT(_sdf->HasElement("max_thrust"), "no max thrust provided");
-  max_thrust_ = _sdf->Get<double>("max_thrust");
+  GZ_ASSERT(_sdf->HasElement("max_torque"), "no max thrust provided");
+  max_torque_ = _sdf->Get<double>("max_torque");
 
   GZ_ASSERT(_sdf->HasElement("max_velocity"), "no max velocity provided");
   max_velocity_ = _sdf->Get<double>("max_velocity");
 
-  base_link_ = model_->GetLink("base_link");
+  back_left_wheel_joint_ = model_->GetJoint("base_link_to_back_left_wheel_link_joint");
+  back_right_wheel_joint_ = model_->GetJoint("base_link_to_back_right_wheel_link_joint");
 }
 
-void BackWheelPlugin::Callback(const indyav_control::ThrustStamped& _msg)
+void BackWheelPlugin::Callback(const indyav_control::TorqueStamped& _msg)
 {
   // TODO: simulate engine input delay
-  if (abs(_msg.thrust) <= max_thrust_)
-    thrust_.X(_msg.thrust);
+  if (abs(_msg.torque) <= max_torque_)
+    torque_ = _msg.torque;
 }
 
 void BackWheelPlugin::OnUpdate()
 {
-  // TODO: only apply forces when wheel collisions are touching the ground
   // TODO: apply forces on base_link at the offsets of the wheels in varrying amounts
   //  to replicate a differential torque thingy
   if (model_->RelativeLinearVel().X() < max_velocity_)
-    base_link_->AddLinkForce(thrust_, ignition::math::Vector3<double>::Zero);
+  {
+    back_left_wheel_joint_->SetVelocity(0, torque_);
+    back_right_wheel_joint_->SetVelocity(0, torque_);
+  }
+  else
+  {
+    back_left_wheel_joint_->SetVelocity(0, max_velocity_);
```

Comment:
What does this do exactly and why?

Reply:
This sets the velocity of the the wheel to a certain speed. By doing so, the begins rotating at a speed corresponding with the number. I am not sure its relationship with the physics though. I will have to look that up.

Reply:
no i mean why are we setting the rotational velocity of the wheel joints to the max_velociy_ (a linear velocity) when the car's lear velocity is >= the max_velocity_

Reply:
Ah! Nice catch